Add documentation

Yoshanuikabundi opened the pull request:
## Description
This PR adds documentation to openff-units. I've already set up RTD so we should see a preview of it in this PR.

This PR adds a 1-page user's guide, as well as some reasonably polished API docs. I've written brief docstrings for everything that was missing them, except inherited methods and attributes that I can't change. I've also declared the public API in `__all__` module attributes, which the docs use to decide what to document.

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [x] API docs
  - [x] Book docs
  - [x] Re-exports for Unit, Quantity and Measurement from `openff.units`

## Questions

 - [ ] Is re-exporting Unit, Quantity and Measurement from `openff.units` as good an idea as I think it is? I'm hoping it just simplifies exploring and memorising the API - now no-one has to remember that `openff.units.units` is a thing.
   - [ ] Should these re-exports be lazy, like the openff.toolkit re-exports?

## Status
- [x] Ready to go

mattwthompson commented on docs/index.md:
```diff
@@ -0,0 +1,159 @@
+
+## Using OpenFF Units
+
+OpenFF Units provides the [`Quantity`] class, which represents a numerical value with units. To create a `Quantity`, multiply a value by the appropriate unit:
+
```
I find myself using the explicit construction
```
mass_proton = unit.Quantity(1.007, unit.amu)
# or even more explicit, maybe TOO explicit
mass_proton = unit.Quantity(value=1.007, units=unit.amu)

```

from time to time because it avoids long expressions on one line, particularly when using arrays and long (i.e. kcal/mol) units.. Including this in the docs would make it a bit more explicit what object users are getting here, i.e. introducing it like this and then "but you can also just use a `*` operator ...." and stick with the shorthand.

If this is too much too fast for an introduction, I'd still like to see something like

```python
>>> # oh by the way, here's the type of the object we made...
>>> type(mass_proton)
openff.units.units.Quantity
```

mattwthompson commented on openff/units/simtk.py:
```diff
@@ -7,6 +11,14 @@
 
 from openff.units import unit
 
+__all__ = [
+    "from_simtk",
+    "to_simtk",
+    "simtk_unit_to_string",
+    "string_to_simtk_unit",
+]
```
Could you remind me the purpose of these `__all__`? I have a distaste for them because they are really easy to de-sync with the module contents and I don't like star imports.

Yoshanuikabundi replied:
They're there to tell Sphinx what parts of the API to document. There are basically four ways to tell Sphinx what to document:

1. Write out everything you want to document in an RST file (like the Toolkit does). This goes out of sync even faster than `__all__`; at least `__all__` can trip up linters and is in the right file.
2. Treat stuff starting with an underscore as private and document everything else, excluding imported items. This means no re-exports which is basically a nonstarter. I've tried a lot of workarounds but Sphinx is stubborn about this, you can rename imports or assign them to variables and it won't document them if imported items are excluded.
3. Treat stuff starting with an underscore as private and document everything else, including imported items. This means you have to rename everything you import or else it gets documented, and the docs get cluttered because people tend to forget to make things private.
4. Document the stuff in `__all__`, which is Python's standard way to declare stuff as public. At least this way the public API that you're promising not to break is opt-in rather than opt-out.

These all suck but 4 sucks the least. For context, 4 is new; I contributed it to Sphinx because 1-3 weren't workable. I'm not saying that makes it good, but it gives some idea of how frustrating this particular wart has been for me :upside_down_face:

mattwthompson reviewed (approved):
Thanks for the overhaul. I might have jumped the gun reviewing this - happy to pass through it again once you see it as complete if there are other things you want feedback on. (I see you've moved your notes into a separate issue and I'll continue those ideas from there.)

I didn't look at the files that I thought might boilerplate; as long as the docs continue to build their purpose is served.

The only thing that looks a bit odd to me is that the user guide stuff is on the right hand side, which is not where I'm used to looking for content (either in our docs or other projects). Was this intentional?

<img width="1500" alt="image" src="https://user-images.githubusercontent.com/7935382/166941707-9fee84a4-4827-424c-984e-628f0879a292.png">

Yoshanuikabundi commented on docs/index.md:
````diff
@@ -0,0 +1,159 @@
+<Quantity(1.0, 'nanometer')>
+```
+
+The underlying value without units can be retrieved with the [`.m`] or [`.magnitude`] properties. Just make sure it's in the units you expect first:
````
Sure thing!

Yoshanuikabundi commented:
> The only thing that looks a bit odd to me is that the user guide stuff is on the right hand side, which is not where I'm used to looking for content (either in our docs or other projects). Was this intentional?

Semi-intentional; the table of contents for the current page goes on the right by default, and the toc for the whole site goes on the left. But you're right, for one-page docs they should go together. The theme has a way to do this but it's not co-operating with me at the moment, so I'll sort it out and fix your other comments on my next working day.

Yoshanuikabundi commented:
Okay, I fixed the TOC on LHS issue, and wrote up your suggestions (thank you!), and I added some explanation of the `to_openmm` and `from_openmm` methods. 

I think this is ready to go but I'd just like to make sure you've noticed that I've added `Measurement`, `Quantity`, and `Unit` as re-exports to the openff.units module. This way we have 
- `openff.units`, the package, with its members 
    - `openff.units.unit` 
    - and `openff.units.Unit`

instead of

- `openff.units`, the package, with 
    - its technically-not-a-module `openff.units.unit`, 
    - its module `openff.units.units`, 
        - with its member `openff.units.units.Unit`

Which I think is a much simpler mental model in terms of remembering where all the things named unit/units go. There's only one `units`, and the purposes of the two `unit` objects are relatively clear from the naming convention. Import times shouldn't be different because `openff.units` already re-exports `unit` from `openff.units.units`

This also means users don't have to worry about the units module, as all its public members are re-exported from the package root. As such, I've excluded this module from the API docs. Let me know if I should put it back.

mattwthompson commented:
> ... re-exports  ...

IIUC, this also means `Measurement` and `Quantity` are top-level?

```python3
>>> from openff.units import *
>>> [key for key in locals() if not key.startswith("_")]
['unit', 'Quantity', 'Measurement', 'Unit']
```

I have a habit of `unit.Quantity(numbers, unit.whatever_unit)` which I'm pretty happy with:

```python3
>>> from openff.units import unit
>>> unit.Quantity(4, unit.angstrom)
<Quantity(4, 'angstrom')>
```

but this approach is a bit cleaner (and much cleaner if making many quantity/measurement objects) if trying to avoid the `unit` registry acting like a namespace:

```python3
>>> from openff.units import unit, Quantity
>>> Quantity(4, unit.angstrom)
<Quantity(4, 'angstrom')>
```

and it works with the pattern I have been using until now. So this seems fine with me and I can't see a downside.

mattwthompson commented:
> Is re-exporting Unit, Quantity and Measurement from openff.units as good an idea as I think it is?

I'm on board with it.

> Should these re-exports be lazy, like the openff.toolkit re-exports?

No preference, I doubt these add much import overhead. Feel free to profile if you wish, but I don't think that's at all necessary. If it's an issue we can deal with it in the future.


Just to be clear - I think this is good to go to. Feel free to merge as you wish or continue to iterate on anything - though everything seems resolved to me. Thanks for the contribution!
